src/twisted/spread/test/test_pb.py

```diff
@@ -1434,19 +1511,34 @@ def test_loginLogout(self):
         # ignore it.
         mind = "BRAINS!"
 
-        d = factory.login(creds, mind)
+        loginCompleted = Deferred()
+
+        d = self.clientFactory.login(creds, mind)
         def cbLogin(perspective):
             self.assertTrue(self.realm.lastPerspective.loggedIn)
             self.assertIsInstance(perspective, pb.RemoteReference)
-            return self._disconnect(None, factory)
+            return loginCompleted
+
+        def cbDisconnect(ignored):
+            self.clientFactory.disconnect()
+            self.completeClientLostConnection()
+
         d.addCallback(cbLogin)
+        d.addCallback(cbDisconnect)
 
         def cbLogout(ignored):
             self.assertTrue(self.realm.lastPerspective.loggedOut)
         d.addCallback(cbLogout)
 
-        connector = reactor.connectTCP("127.0.0.1", self.portno, factory)
-        self.addCleanup(connector.disconnect)
+        self.establishClientAndServer()
+        self.pump.flush()
+        # The perspective passed to cbLogin has gone out of scope.
+        # Ensure its __del__ runs...
+        gc.collect()
+        # ...and send its decref message to the server
+        self.pump.flush()
+        # Now allow the client to disconnect.
+        loginCompleted.callback(None)
         return d
```

Comment:
Synchronous tests should no longer return `Deferred`s.  This is what tells Twisted to run the main loop.